src/main/java/org/prebid/server/bidder/msft/MsftBidder.java

```diff
@@ -0,0 +1,403 @@
+                .map(ExtApp::getPrebid);
+
+        final String source = prebid.map(ExtAppPrebid::getSource).orElse(null);
+        final String version = prebid.map(ExtAppPrebid::getVersion).orElse(null);
+
+        return ObjectUtils.allNotNull(source, version)
+                ? "%s-%s".formatted(source, version)
+                : StringUtils.EMPTY;
+    }
+
+    private ExtImpMsft parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), MSFT_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException("Failed to deserialize Microsoft imp extension: " + e.getMessage());
+        }
+    }
+
+    private Imp updateImp(Imp imp, ExtImpMsft extImp, String defaultDisplayManagerVer) {
+        final Imp.ImpBuilder impBuilder = imp.toBuilder();
+
+        if (StringUtils.isNotEmpty(extImp.getInvCode())) {
+            impBuilder.tagid(extImp.getInvCode());
+        }
+
+        if (imp.getBanner() != null) {
+            impBuilder.banner(updateBanner(imp.getBanner(), extImp));
+        }
+
+        if (StringUtils.isEmpty(imp.getDisplaymanagerver())) {
+            impBuilder.displaymanagerver(defaultDisplayManagerVer);
+        }
+
+        impBuilder.ext(updateImpExt(imp, extImp));
+
+        return impBuilder.build();
+    }
+
+    private Banner updateBanner(Banner banner, ExtImpMsft extImp) {
+        final Banner.BannerBuilder bannerBuilder = banner.toBuilder();
+
+        final List<Format> bannerFormat = banner.getFormat();
+        if (banner.getW() == null && banner.getH() == null && CollectionUtils.isNotEmpty(bannerFormat)) {
+            final Format firstFormat = bannerFormat.getFirst();
+            bannerBuilder.w(firstFormat.getW());
+            bannerBuilder.h(firstFormat.getH());
+        }
+
+        if (banner.getApi() == null) {
+            bannerBuilder.api(extImp.getBannerFrameworks());
+        }
+
+        return bannerBuilder.build();
+    }
+
+    private ObjectNode updateImpExt(Imp imp, ExtImpMsft extImp) {
+        final ObjectNode updatedImpExt = mapper.mapper().createObjectNode()
+                .set("appnexus", mapper.mapper().valueToTree(
+                        MsftExtImpOutgoing.builder()
+                                .placementId(extImp.getPlacementId())
+                                .allowSmallerSizes(extImp.getAllowSmallerSizes())
+                                .usePmtRule(extImp.getUsePmtRule())
+                                .keywords(extImp.getKeywords())
+                                .trafficSourceCode(extImp.getTrafficSourceCode())
+                                .pubClick(extImp.getPubclick())
+                                .extInvCode(extImp.getExtInvCode())
+                                .extImpId(extImp.getExtImpId())
+                                .build()));
+
+        final String gpid = ObjectUtil.getIfNotNull(imp.getExt().get("gpid"), JsonNode::textValue);
```

Comment:
Use `.at()` to simplify. This will return `MissingNode` instead of null in case of absence, which `.asText()` will return empty string.